// buc-utils/index.js
/**!
 * buc-utils - index.js
 *
 * Copyright(c) Alibaba Group Holding Limited.
 *
 * Authors:
 *   苏千 <suqian.yf@alipay.com> (http://fengmk2.github.com)
 */

'use strict';

/**
 * Module dependencies.
 */

var copyto = require('copy-to');

exports.formatUser = function (data) {
  var name = data.nickNameCn || data.lastName;
  var email = data.emailAddr;
  var emailPrefix = data.emailPrefix;
  if (!email && data.account) {
    email = data.account;
    if (email.indexOf('@') === -1) {
      email += '@alibaba-inc.com';
    }
  }

  if (!emailPrefix && email) {
    emailPrefix = email.split('@', 1)[0];
  }

  var user = {
    // buc id
    bucid: data.id,
    // 域账号
    login: emailPrefix,
    // 主邮箱地址
    email: email,
    // 优先使用花名
    name: name,
    // 阿里内网个人页面
    html_url: 'https://work.alibaba-inc.com/work/u/' + data.empId,
    // 内网头像
    avatar_url: 'https://work.alibaba-inc.com/photo/' + data.empId + '.220x220.jpg',
    // 旺旺链接
    im_url: 'https://work.alibaba-inc.com/new_amos_im_alisoft_com/msg.aw?v=2&uid=' +
      encodeURIComponent(name) + '&site=cntaobao&charset=utf-8&s=1'
  };
  copyto(data).to(user);
  return user;
};
